### PR TITLE
migration from stable8.2 to stable9

### PR DESCRIPTION
**Pull request by phisch**
## Description
This PR adds a migration from stable8.2 to stable9. The migration runs before the stable9 to stable10 migration, therefore it is possible to upgrade from stable8.2 to stable10 directly.

## Related Issue
https://github.com/owncloud/enterprise/issues/2010

## Motivation and Context
So people can update more easily.

## How Has This Been Tested?
Currently only by hand, by me. Will be tested automatically. Need help from qa though!

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation. <- maybe, to get awareness
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/Migrations/Version20170101010100.php [hunk 1]
```diff
@@ -0,0 +1,1099 @@
...
+					'unsigned' => true
+				]
+			);
+
+			$systemTagObjectMappingTable->addUniqueIndex(['objecttype', 'objectid', 'systemtagid'], 'mapping');
+
+		}
+
+		if ($schema->hasTable("${prefix}file_map")) {
+			$schema->dropTable("${prefix}file_map");
```
Comment (phisch) [inline core/Migrations/Version20170101010100.php]: @DeepDiver1975 @PVince81 i don't know what data is in this table. Is it important, do i have to transform and put it into another table?

Comment (DeepDiver1975) [inline core/Migrations/Version20170101010100.php]: no - this data is not needed

Comment (PVince81) [inline core/Migrations/Version20170101010100.php]: :fire:

Comment (phisch) [inline core/Migrations/Version20170101010100.php]: :+1:

**Comment by PVince81** [comment]
In the past, some apps had update routines in "apps/$appName/appinfo/update.php". We need to make sure these still exist and run on stable10.

The ones I know of that I checked:
- [x] files: seems ok
- [ ] 🚫 files_trashbin: file exists but some pieces were deleted in stable10
- [x] files_sharing update routine to flatten shares: update.php is still there
- [ ] 🚫 files_external update routine is gone: it converts an older storage config format to "oc_mounts" etc tables
- [ ] 🚫 files_versions: file deleted in stable10
- [ ] encryption...

- [ ] To be thorough we should also check whether stable9 and stable9.1 had update files/code that might also have been removed.

**Comment by PVince81** [comment]
One worry is that some "update.php" code relies on other stuff being migrated first through other means.

So if we put these update.php routines into migrations, I suggest to order them exactly by order of major versions. So run the update.php codes from stable9 before running the stable9.1 migrations, etc. Just to be sure.

Ideal would be to also have test cases that cover whatever these update.php routines were fixing.

**Comment by phisch** [comment]
> files_trashbin: file exists but some pieces were deleted in stable10
The expire trash background task has been moved into the `appinfo.xml`: https://github.com/owncloud/core/blob/master/apps/files_trashbin/appinfo/info.xml

All fine with that one!

**Comment by phisch** [comment]
> files_external update routine is gone: it converts an older storage config format to "oc_mounts" etc tables

This is a real concern. We need to add the migration steps that are here:
https://github.com/owncloud/core/blob/stable9/apps/files_external/migration/storagemigrator.php

@PVince81 should i add it to the app again? Or should we include it in the core migrations?

**Comment by phisch** [comment]
> files_versions: file deleted in stable10

Background available as Command in https://github.com/owncloud/core/blob/master/apps/files_versions/appinfo/info.xml

**Comment by PVince81** [comment]
if adding stuff back to update.php is quicker then I suggest we go that route. I expect that after the OC to 10.0 migration we might remove these again at some point...

**Review by DeepDiver1975 — APPROVED** [review]

**Comment by PVince81** [comment]
I'd like to see the checkboxes here updated before merging: https://github.com/owncloud/core/pull/28256#issuecomment-311898504

**Comment by phisch** [comment]
The storagemigration (https://github.com/owncloud/core/blob/stable9/apps/files_external/migration/storagemigrator.php) can't be added here because of missing services in the files_external app.

If we want to support migrating global and user mounts, we should add those legacy services again to files_external and add it as migration there.

**Comment by PVince81** [comment]
> If we want to support migrating global and user mounts, we should add those legacy services again to files_external and add it as migration there.

that would be fine

**Comment by PVince81** [comment]
@phisch can you post a list of items/checkbox of the ton of testing you have done lately?

**Comment by PVince81** [comment]
and also please update the checkboxes above with the bits you have already fixed. Thanks

**Comment by phisch** [comment]
Tested the migration for all major databases:
 - [x] sqlite
 - [x] mysql
 - [x] pgsql
 - [x] oracle

Looked for missing actions/migrations from `update.php` files in apps. There are no missing actions except for the mounts that will be added in the `files_external` app.

Merging this PR.

**Comment by PVince81** [comment]
@phisch thanks. Will you add this separately ?

> add migrations for https://github.com/owncloud/core/blob/stable9/apps/files_external/migration/storagemigrator.php

**Comment by PVince81** [comment]
please backport this PR to stable10

**Comment by phisch** [comment]
> Will you add this separately ?
@PVince81 yes

Backport incomming.

**Pull request merged**